libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1707,19 +1708,64 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
-				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
-				}
+				let spill := sub(length, mul(fullSlots, <itemsPerSlot>))
+
+				<?sameStride>
+					let i := 0
+					for { } lt(i, fullSlots) { i := add(i, 1) } {
+						sstore(add(dstSlot, i), <maskFull>(sload(add(srcSlot, i))))
+					}
+					if gt(spill, 0) {
+						sstore(add(dstSlot, i), <maskBytes>(sload(add(srcSlot, i)), mul(spill, <dstStride>)))
+					}
+				<!sameStride>
+					let srcSlotValue := sload(srcSlot)
+					let srcItemIndexInSlot := 0
+					let totalSlots := fullSlots
+					if gt(spill, 0) {
+						totalSlots := add(totalSlots, 1)
+					}
+					for { let i := 0 } lt(i, totalSlots) { i := add(i, 1) } {
+						let dstSlotValue := 0
+						let itemsInSlot := <itemsPerSlot>
+						if eq(i, fullSlots) { itemsInSlot := spill }
+						for { let j := 0 } lt(j, itemsInSlot) { j := add(j, 1) } {
+							let itemValue := <convert>(
+								<?leftAligned>
+									<srcAlignLeft>(and(srcSlotValue, <srcMask>)))
+								<!leftAligned>
+									and(srcSlotValue, <srcMask>))
+								</leftAligned>
+							dstSlotValue := or(
+								dstSlotValue,
+								<?leftAligned>
+									<shiftRightDyn>(sub(256, mul(add(j, 1), mul(<dstStride>, 8))), itemValue)
+								<!leftAligned>
+									<shiftLeftDyn>(mul(j, mul(<dstStride>, 8)), itemValue)
```

Comment:
I think I can get rid of dynamic shifts here, and will try that now

Reply:
It is not as easy as I was thinking. There is going to be at least one dynamic shift call per source slot...
So, not sure what will be the difference in gas costs, is it worth optimizing this one?

Reply:
Let's do that in the optimizer.